### PR TITLE
#43 - Add interfaces for content models of Shopware 6

### PR DESCRIPTION
**Pull request by rmakara**
**Changes:**
* Add interfaces for `content` group based on Shopware 6 data structures.
* These models seem to be required for covering Shopware API endpoints that we will need.
* It creates the following structure below `content`:
  * category
  * cms
  * mail-template
  * media
  * newsletter
  * product
  * product-stream
  * property
  * rule

**Issue:**
#43

### CODE REVIEW

File: packages/shopware-6-client/src/interfaces/models/content/category/Category.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  translations: CategoryTranslation[] | null;
+  media: Media | null;
+  products: Product[] | null;
+  nestedProducts: Product[] | null;
+  afterCategoryId: string | null;
+  customFields: CustomField[];
+  tags: TagEntity[] | null;
+  cmsPageId: string | null;
+  cmsPage: CmsPage | null;
+  slotConfig: [] | null;
```
Comment (patzick) [inline packages/shopware-6-client/src/interfaces/models/content/category/Category.ts]: missing type

File: packages/shopware-6-client/src/interfaces/models/content/mail-template/MailTemplateType.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import { MailTemplateType } from "./MailTemplateType";
+import { MailTemplateSalesChannel } from "./MailTemplateSalesChannel";
+import { MailTemplateTypeTranslation } from "./MailTemplateTypeTranslation";
+import { CustomField } from "../../common/CustomField";
+export interface MailTemplateType {
+  name: string;
+  technicalName: string;
+  availableEntities: [] | null;
```
Comment (patzick) [inline packages/shopware-6-client/src/interfaces/models/content/mail-template/MailTemplateType.ts]: missing type

File: packages/shopware-6-client/src/interfaces/models/content/media/Media.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+import { User } from "../../system/user/User";
+
+export interface Media {
+  userId: string | null;
+  mimeType: string | null;
+  fileExtension: string | null;
+  fileSize: number | null;
+  title: string | null;
+  metaDataRaw: string | null;
+  metaData: [] | null;
```
Comment (patzick) [inline packages/shopware-6-client/src/interfaces/models/content/media/Media.ts]: missing type

File: packages/shopware-6-client/src/interfaces/models/content/product-stream/ProductStream.ts [hunk 4]
```diff
@@ -0,0 +1,12 @@
+import { ProductStreamTranslation } from "./ProductStreamTranslation";
+import { ProductStreamFilter } from "./ProductStreamFilter";
+import { CustomField } from "../../common/CustomField";
+export interface ProductStream {
+  name: string;
+  description: string | null;
+  apiFilter: [] | null;
```
Comment (patzick) [inline packages/shopware-6-client/src/interfaces/models/content/product-stream/ProductStream.ts]: missing type

File: packages/shopware-6-client/src/interfaces/models/content/product-stream/ProductStreamFilter.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+  field: string | null;
+  operator: string | null;
+  value: string | null;
+  productStreamId: string;
+  parentId: string | null;
+  productStream: ProductStream | null;
+  queries: ProductStreamFilter[] | null;
+  parent: ProductStreamFilter | null;
+  position: number;
+  parameters: [] | null;
```
Comment (patzick) [inline packages/shopware-6-client/src/interfaces/models/content/product-stream/ProductStreamFilter.ts]: missing type

File: packages/shopware-6-client/src/interfaces/models/content/product/Product.ts [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+  purchasePrice: number | null;
+  markAsTopseller: boolean | null;
+  weight: number | null;
+  width: number | null;
+  height: number | null;
+  length: number | null;
+  releaseDate: Date;
+  categoryTree: [] | null;
+  optionsIds: [] | null;
+  propertyIds: [] | null;
```
Comment (patzick) [inline packages/shopware-6-client/src/interfaces/models/content/product/Product.ts]: missing types

File: packages/shopware-6-client/src/interfaces/models/content/product/Product.ts [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+  media: ProductMedia[];
+  translations: ProductTranslation[];
+  categories: Category[];
+  tags: TagEntity[];
+  properties: PropertyGroupOption[] | null;
+  options: PropertyGroupOption[] | null;
+  categoriesRo: Category[] | null;
+  coverId: string | null;
+  customFields: CustomField[];
+  tagIds: [] | null;
```
Comment (patzick) [inline packages/shopware-6-client/src/interfaces/models/content/product/Product.ts]: missing type

File: packages/shopware-6-client/src/interfaces/models/content/product/Product.ts [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+  productReviews: ProductReview[] | null;
+  ratingAverage: number | null;
+  extensions: [];
```
Comment (patzick) [inline packages/shopware-6-client/src/interfaces/models/content/product/Product.ts]: missing type

File: packages/shopware-6-client/src/interfaces/models/content/product/ProductConfiguratorSetting.ts [hunk 9]
```diff
@@ -0,0 +1,16 @@
+import { Media } from "../media/Media";
+import { PropertyGroupOption } from "../property/PropertyGroupOption";
+import { Product } from "./Product";
+import { CustomField } from "../../common/CustomField";
+export interface ProductConfiguratorSetting {
+  productId: number;
+  optionId: number;
+  mediaId: number;
+  position: number;
+  price: [] | null;
```
Comment (patzick) [inline packages/shopware-6-client/src/interfaces/models/content/product/ProductConfiguratorSetting.ts]: missing type

File: packages/shopware-6-client/src/interfaces/models/content/rule/Rule.ts [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+import { PromotionSetGroup } from "../../checkout/promotion/PromotionSetGroup";
+import { Promotion } from "../../checkout/promotion/Promotion";
+import { RuleCondition } from "./RuleCondition";
+import { PaymentMethod } from "../../checkout/payment/PaymentMethod";
+import { CustomField } from "../../common/CustomField";
+export interface Rule {
+  name: string;
+  description: string | null;
+  payload: string | Rule | null;
+  moduleTypes: [] | null;
```
Comment (patzick) [inline packages/shopware-6-client/src/interfaces/models/content/rule/Rule.ts]: missing type

File: packages/shopware-6-client/src/interfaces/models/content/rule/RuleCondition.ts [hunk 11]
```diff
@@ -0,0 +1,14 @@
+import { Rule } from "./Rule";
+import { RuleCondition } from "./RuleCondition";
+import { CustomField } from "../../common/CustomField";
+export interface RuleCondition {
+  type: string;
+  ruleId: string;
+  parentId: string | null;
+  value: [] | null;
```
Comment (patzick) [inline packages/shopware-6-client/src/interfaces/models/content/rule/RuleCondition.ts]: missing type

File: packages/shopware-6-client/src/interfaces/models/content/category/Category.ts [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+  externalLink: string | null;
+  visible: boolean;
+  type: string;
+  description: string;
+  extensions: [];
```
Comment (patzick) [inline packages/shopware-6-client/src/interfaces/models/content/category/Category.ts]: missing type

File: packages/shopware-6-client/src/interfaces/models/content/newsletter/NewsletterRecipient.ts [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+  salutation: Salutation | null;
+  languageId: string;
+  language: Language | null;
+  salesChannelId: string;
+  salesChannel: SalesChannel | null;
+  customFields: CustomField[];
+  confirmedAt: Date | null;
+  createdAt: Date;
+  updatedAt: Date | null;
+  tags: TagEntity[] | null;
```
Comment (patzick) [inline packages/shopware-6-client/src/interfaces/models/content/newsletter/NewsletterRecipient.ts]: please rename to Tag

File: packages/shopware-6-client/src/interfaces/models/content/category/Category.ts (hunk 1 above)
Comment (rmakara) [reply packages/shopware-6-client/src/interfaces/models/content/category/Category.ts]: Just an array in Shopware.

File: packages/shopware-6-client/src/interfaces/models/content/mail-template/MailTemplateType.ts (hunk 2 above)
Comment (rmakara) [reply packages/shopware-6-client/src/interfaces/models/content/mail-template/MailTemplateType.ts]: Just an array in Shopware.

File: packages/shopware-6-client/src/interfaces/models/content/media/Media.ts (hunk 3 above)
Comment (rmakara) [reply packages/shopware-6-client/src/interfaces/models/content/media/Media.ts]: Just an array in Shopware.

File: packages/shopware-6-client/src/interfaces/models/content/product-stream/ProductStream.ts (hunk 4 above)
Comment (rmakara) [reply packages/shopware-6-client/src/interfaces/models/content/product-stream/ProductStream.ts]: Just an array in Shopware.

File: packages/shopware-6-client/src/interfaces/models/content/product-stream/ProductStreamFilter.ts (hunk 5 above)
Comment (rmakara) [reply packages/shopware-6-client/src/interfaces/models/content/product-stream/ProductStreamFilter.ts]: Just an array in Shopware.

File: packages/shopware-6-client/src/interfaces/models/content/product/ProductConfiguratorSetting.ts (hunk 9 above)
Comment (rmakara) [reply packages/shopware-6-client/src/interfaces/models/content/product/ProductConfiguratorSetting.ts]: Just an array in Shopware.

File: packages/shopware-6-client/src/interfaces/models/content/rule/RuleCondition.ts (hunk 11 above)
Comment (rmakara) [reply packages/shopware-6-client/src/interfaces/models/content/rule/RuleCondition.ts]: Just an array in Shopware.

File: packages/shopware-6-client/src/interfaces/models/content/category/Category.ts (hunk 12 above)
Comment (rmakara) [reply packages/shopware-6-client/src/interfaces/models/content/category/Category.ts]: Hm, that array comes from very deeply inherited trait, which is `ExtendableTrait.php`. I believe if we go through all the interfaces we would need to add that to most of the interfaces.

I will delete that field for now and we will consider adding that to `ExtendableTrait`, which will be extended by `Entity`, which will be extended by concrete interfaces.

Deleting it for now as we did not include it in all other interfaces.

File: packages/shopware-6-client/src/interfaces/models/content/newsletter/NewsletterRecipient.ts (hunk 13 above)
Comment (rmakara) [reply packages/shopware-6-client/src/interfaces/models/content/newsletter/NewsletterRecipient.ts]: Done, sir! :)

File: packages/shopware-6-client/src/interfaces/models/content/rule/Rule.ts (hunk 10 above)
Comment (rmakara) [reply packages/shopware-6-client/src/interfaces/models/content/rule/Rule.ts]: Just an array in Shopware.

**Comment by rmakara** [comment]
@patzick all resolved. :)

Longer story about `Category.extensions` field because it comes from deeply added trait. I deleted this now instead of adding a type for that.

**Pull request merged**